experimental/JuLie/partitions.jl

```diff
@@ -0,0 +1,727 @@
+  k >= 0 || throw(ArgumentError("k >= 0 required"))
+  n = ZZ(n)
+  k = ZZ(k)
+
+  # Special cases
+  if n == k
+    return ZZ(1)
+  elseif n < k || k == 0
+    return ZZ(0)
+  elseif k == 1
+    return ZZ(1)
+
+    # See https://oeis.org/A008284
+  elseif n < 2*k
+    return num_partitions(n-k) #n-k>=0 holds since the case n<k was already handled
+
+    # See https://oeis.org/A008284
+  elseif n <= 2+3*k
+    p = num_partitions(n-k) #n-k>=0 holds since the case n<k was already handled
+    for i=0:Int(n)-2*Int(k)-1
+      p = p - num_partitions(ZZ(i))
+    end
+    return p
+
+    # Otherwise, use recurrence
+    # The following is taken from the GAP code in lib/combinat.gi
+    # It uses the standard recurrence relation but in a more intelligent
+    # way without recursion.
+  else
+    n = Int(n)
+    k = Int(k)
+    p = fill( ZZ(1), n )
+    for l = 2:k
+      for m = l+1:n-l+1
+        p[m] = p[m] + p[m-l]
+      end
+    end
+    return p[n-k+1]
+  end
+
+end
+
+
+@Markdown.doc """
+    partitions(n::IntegerUnion)
+
+A list of all partitions of an integer ``n ≥ 0``, produced in
+lexicographically *descending* order. This ordering is like in Sage, but
+opposite to GAP. You can apply the function `reverse` to reverse the
+order. As usual, you may increase performance by using smaller integer types.
+The algorithm used is "Algorithm ZS1" by Zoghbi & Stojmenovic (1998); see
+[ZS98](@cite).
+
+# Examples
+```jldoctest
+julia> partitions(Int8(4))
+5-element Vector{Partition{Int8}}:
+ Int8[4]
+ Int8[3, 1]
+ Int8[2, 2]
+ Int8[2, 1, 1]
+ Int8[1, 1, 1, 1]
+```
+"""
+function partitions(n::IntegerUnion)
+
+  #Argument checking
+  n >= 0 || throw(ArgumentError("n >= 0 required"))
+
+  # Use type of n
+  T = typeof(n)
+
+  # Some trivial cases
+  if n==0
+    return Partition{T}[ Partition{T}([]) ]
+  elseif n==1
+    return Partition{T}[ Partition{T}([1]) ]
+  end
+
+  # Now, the algorithm starts
+  P = Partition{T}[]    #this will be the array of all partitions
+  k = 1
+  q = 1
+  d = fill( T(1), n )
+  d[1] = n
+  push!(P, Partition{T}(d[1:1]))
+  while q != 0
+    if d[q] == 2
+      k += 1
+      d[q] = 1
+      q -= 1
+    else
+      m = d[q] - 1
+      np = k - q + 1
+      d[q] = m
+      while np >= m
+        q += 1
+        d[q] = m
+        np = np - m
+      end
+      if np == 0
+        k = q
+      else
+        k = q + 1
+        if np > 1
+          q += 1
+          d[q] = np
+        end
+      end
+    end
+    push!(P, Partition{T}(d[1:k]))
+  end
+  return P
+
+end
+
+
+@Markdown.doc """
+    ascending_partitions(n::IntegerUnion;alg="ks")
+
+Instead of encoding a partition of an integer ``n ≥ 0`` as a *descending*
+sequence (which is our convention), one can also encode it as an *ascending*
+sequence. In the papers Kelleher & O'Sullivan (2014) and Merca (2012) it is
+said that generating the list of all ascending partitions is more efficient
+than generating descending ones. To test this, I have implemented the
```

Comment:
This docstring uses "I" quite a bit: "I have", "I am", .... For an OSCAR user it won't be clear to whom that "I" refers. So we should rewrite it, either using "we"; or using "U. Thiel"

@ulthiel any preferences?

Reply:
Well, this was my personal documentation, so of course I used "I". I don't know. It's maybe also not so important.

Reply:
Noticing that the ascending_partitions stuff is in the pull request as well: I later changed this to ascending_compositions (which was part of my composition generator) because it's technically not a partition by definition. Since the composition stuff is currently not in the pull request (and not in JuLie) I'm unsure what should happen with this functionality. It should definitely be there eventually.

Reply:
Of course it made perfect sense to use "I" while this was in JuLie, and I am not criticizing this. I am merely pointing out that it doesn't make sense anymore in the context of Oscar, because most readers won't know whom "I" refers to. Moreover, it differs from the style we used elsewhere. Hence it should be changed. I can easily change it now, of course, but as a courtesy I thought I'd ask you for your preferences...?

Regarding your second reply: Note that this is the docstring for `ascending_partitions` we are discussing here. So you are saying you renamed it later anyway, and moved it to some other module... So should we just delete it from this PR for now and it'll be added back later?